Ensure legacy sparse constructor/new doesn't interpret python data as tensor data.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31490 Ensure legacy sparse constructor/new doesn't interpret python data as tensor data.**
* #31485 Error out if legacy Tensor.new is called on alternate layouts / dtypes

When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

Differential Revision: [D19196498](https://our.internmc.facebook.com/intern/diff/D19196498)

gchanan committed:
Ensure legacy sparse constructor/new doesn't interpret python data as tensor data.

When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 8dfa3b3:


* **2/2** failures introduced in this PR
* **2/2** recognized as flaky
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8dfa3b3fe40f29e64d86574dae06e3d992b497b8).

### :detective: 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4020297) pytorch_windows_test2 (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77678067) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77678067))
<details>
<summary>
<code> - cmake (exited 1) - cmake not installed. An error occurred during installation:</code>
</summary>

```
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
cmake package files install completed. Performing other installation steps. 
The install of cmake was NOT successful. 
cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
 
Chocolatey installed 0/1 packages. 1 packages failed. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log). 
 
Failures 
 - cmake (exited 1) - cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4020298) pytorch_windows_test1 (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=77678117) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=77678117))
<details>
<summary>
<code> - cmake (exited 1) - cmake not installed. An error occurred during installation:</code>
</summary>

```
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
cmake package files install completed. Performing other installation steps. 
The install of cmake was NOT successful. 
cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
 
Chocolatey installed 0/1 packages. 1 packages failed. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log). 
 
Failures 
 - cmake (exited 1) - cmake not installed. An error occurred during installation: 
 The remote server returned an error: (503) Server Unavailable. Service Unavailable 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -436,7 +437,8 @@ Tensor legacy_sparse_tensor_new(c10::TensorTypeId type_id, at::ScalarType scalar
       // new(sequence) binds to this signature but should be treated differently
       // unless the sequences is a torch.Size
-      return legacy_new_from_sequence(type_id, scalar_type, deviceOptional, r.pyobject(0));
+      throw TypeError("SparseTensor.new does not support interpreting Python data as tensor data. " \
```
This might not be a very clear error message (I had to read the issue description to understand).
Is the problem that `legacy_new_from_sequence` can only create Tensors with strided layout? Then maybe we want to either:
- Add a layout argument to `legacy_new_from_sequence` to support this properly.
- Change the error message to something like "SparseTensor.new(sequence) only accepts sizes. For sequence containing values, please use the new constructor torch.sparse_coo_tensor() instead.".

gchanan replied:
> Is the problem that legacy_new_from_sequence can only create Tensors with strided layout?

We don't support creating sparse tensors from dense (non-Tensor) data in any constructor.

> Add a layout argument to legacy_new_from_sequence to support this properly.

We shouldn't add functionality to legacy construcotrs.

> Change the error message to something like "SparseTensor.new(sequence) only accepts sizes. For sequence containing values, please use the new constructor torch.sparse_coo_tensor() instead.".

I like this error message better, thanks!

gchanan replied:
ok changed it to:

"SparseTensor.new(sequence) only accepts sizes.  Please use torch.sparse_coo_tensor() or construct a strided tensor and convert it to sparse via to_sparse."

gchanan committed:
Update on "Ensure legacy sparse constructor/new doesn't interpret python data as tensor data."


When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

[ghstack-poisoned]

gchanan committed:
Update on "Ensure legacy sparse constructor/new doesn't interpret python data as tensor data."


When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

[ghstack-poisoned]

gchanan committed:
Update on "Ensure legacy sparse constructor/new doesn't interpret python data as tensor data."


When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

Differential Revision: [D19196498](https://our.internmc.facebook.com/intern/diff/D19196498)

[ghstack-poisoned]

gchanan committed:
Update on "Ensure legacy sparse constructor/new doesn't interpret python data as tensor data."


When this happens, a dense tensor is constructed from a sparse constructor.

Fixes: https://github.com/pytorch/pytorch/issues/16154

Differential Revision: [D19196498](https://our.internmc.facebook.com/intern/diff/D19196498)

[ghstack-poisoned]
